File: code/jvm/domain/src/main/kotlin/ChannelInvitation.kt
import java.time.LocalDateTime

class ChannelInvitation(
    id: Int,
    sender: User,
    val receiver: User,
    val channel: Channel,
    val role: Role,
    isUsed: Boolean = false,
    timestamp: LocalDateTime,
) : Invitation(id, sender, isUsed, timestamp) {
    init {
        require(id >= 0) { "id must be greater than 0" }
        require(role in Role.entries.toTypedArray()) { "Invalid role" }
        require(sender != receiver) { "Sender and receiver must be different" }
        require(timestamp <= LocalDateTime.now()) { "Invalid timestamp" }
    }

    private fun copy() = ChannelInvitation(id, sender, receiver, channel, role, true, timestamp)

    override fun markAsUsed() = copy()
}

File: code/js/src/domain/ChannelInvitation.ts
import { Channel } from "./Channel";
import { Role } from "./Role";
import { User } from "./User";



export type ChannelInvitation = {
    id: number,
    sender: User,
    receiver: User,
    channel: Channel,
    role: Role,
    isUsed: boolean,
    timestamp: Date,
}